### PR TITLE
Refactor P&ID function

### PR DESCRIPTION
**Pull request by haakonvt**
# Description

- Updated to adhere to Functions specifications.
- Split into several files
- Refactored a lot of functionality into smaller functions
- General code modernization and maintenance

### CODE REVIEW

File: cognite_toolkit/cognite_modules/examples/cdf_data_pipeline_files_valhall/functions/fu_context_files_oid_fileshare_annotation/pipeline.py [hunk 1]
```diff
@@ -0,0 +1,554 @@
...
+                append_asset_entities(entities, client, asset_root_xid)
+                annotated_count, error_count = process_files(
+                    client, entities, files_to_process, annotation_list, config
+                )
+            msg = (
+                f"Annotated P&ID files for asset: {asset_root_xid} number of files annotated: {annotated_count}, "
+                f"file not annotaded due to errors: {error_count}"
+            )
+            print(f"[INFO] {msg}")
+            update_extpipe_run(config.extpipe_xid, "success", msg)
```
Comment (doctrino) [inline cognite_toolkit/cognite_modules/examples/cdf_data_pipeline_files_valhall/functions/fu_context_files_oid_fileshare_annotation/pipeline.py:92]:
```suggestion
            update_extpipe_run(client, config.extpipe_xid, "success", msg)
```

File: cognite_toolkit/cognite_modules/examples/cdf_data_pipeline_files_valhall/functions/fu_context_files_oid_fileshare_annotation/pipeline.py [hunk 2]
```diff
@@ -0,0 +1,554 @@
...
+
+        except Exception as e:
+            msg = (
+                f"Annotated P&ID files failed on root asset: {asset_root_xid}. "
+                f"Message: {e!s}, traceback:\n{traceback.format_exc()}"
+            )
+            print(f"[ERROR] {msg}")
+            update_extpipe_run(config.extpipe_xid, "failure", shorten(msg, 1000))
```
Comment (doctrino) [inline cognite_toolkit/cognite_modules/examples/cdf_data_pipeline_files_valhall/functions/fu_context_files_oid_fileshare_annotation/pipeline.py:100]:
```suggestion
            update_extpipe_run(client, config.extpipe_xid, "failure", shorten(msg, 1000))
```

File: cognite_toolkit/cognite_modules/examples/cdf_data_pipeline_files_valhall/functions/fu_context_files_oid_fileshare_annotation/pipeline.py [hunk 3]
```diff
@@ -0,0 +1,554 @@
...
+    client: CogniteClient,
+    asset_root_xid: str,
+    config: AnnotationConfig,
+) -> tuple[dict[str, FileMetadata], dict[str, FileMetadata]]:
+    """
+    Read files based on doc_type and mime_type to find P&ID files
+
+    :returns: dict of files
+    """
+    doc_count = 0
```
Comment (doctrino) [inline cognite_toolkit/cognite_modules/examples/cdf_data_pipeline_files_valhall/functions/fu_context_files_oid_fileshare_annotation/pipeline.py:129]: Not used
```suggestion
```

File: cognite_toolkit/cognite_modules/examples/cdf_data_pipeline_files_valhall/functions/fu_context_files_oid_fileshare_annotation/pipeline.py [hunk 4]
```diff
@@ -0,0 +1,554 @@
...
+    all_pnid_files: dict[str, FileMetadata] = {}
+    pnids_to_process: dict[str, FileMetadata] = {}
+    meta_file_update: list[FileMetadataUpdate] = []
+    print(
+        f"[INFO] Get files to annotate data set: {config.doc_data_set_xid}, asset root: {asset_root_xid} "
+        f"doc_type: {config.pnid_doc_type} and mime_type: {ORG_MIME_TYPE}"
+    )
+    file_list = get_file_list(client, asset_root_xid, config)
+    for file in file_list:
+        doc_count += 1
```
Comment (doctrino) [inline cognite_toolkit/cognite_modules/examples/cdf_data_pipeline_files_valhall/functions/fu_context_files_oid_fileshare_annotation/pipeline.py:139]:
```suggestion
```

File: cognite_toolkit/cognite_modules/examples/cdf_data_pipeline_files_valhall/functions/fu_context_files_oid_fileshare_annotation/pipeline.py [hunk 5]
```diff
@@ -0,0 +1,554 @@
...
+
+def get_files_entities(pnid_files: dict[str, FileMetadata]) -> list[Entity]:
+    """
+    Loop found P&ID files and create a list of entities used for matching against file names in P&ID
+
+    Args:
+        pnid_files: Dict of files found based on filter
+    """
+    entities: list[Entity] = []
+    doc_count = 0
```
Comment (doctrino) [inline cognite_toolkit/cognite_modules/examples/cdf_data_pipeline_files_valhall/functions/fu_context_files_oid_fileshare_annotation/pipeline.py:170]:
```suggestion
```

File: cognite_toolkit/cognite_modules/examples/cdf_data_pipeline_files_valhall/functions/fu_context_files_oid_fileshare_annotation/pipeline.py [hunk 6]
```diff
@@ -0,0 +1,554 @@
...
+
+    for file_xid, file_meta in pnid_files.items():
+        doc_count += 1
```
Comment (doctrino) [inline cognite_toolkit/cognite_modules/examples/cdf_data_pipeline_files_valhall/functions/fu_context_files_oid_fileshare_annotation/pipeline.py:173]:
```suggestion
```

File: cognite_toolkit/cognite_modules/examples/cdf_data_pipeline_files_valhall/functions/fu_context_files_oid_fileshare_annotation/pipeline.py [hunk 7]
```diff
@@ -0,0 +1,554 @@
...
+def get_existing_annotations(client: CogniteClient, entities: list[Entity]) -> dict[Optional[int], list[Optional[int]]]:
+    """
+    Read list of already annotated files and get corresponding annotations
+
+    :param client: Dict of files found based on filter
+    :param entities:
+
+    :returns: dictionary of annotations
+    """
+    annotation_list = AnnotationList([])
```
Comment (doctrino) [inline cognite_toolkit/cognite_modules/examples/cdf_data_pipeline_files_valhall/functions/fu_context_files_oid_fileshare_annotation/pipeline.py:232]: Overwritten below
```suggestion
```

File: cognite_toolkit/cognite_modules/examples/cdf_data_pipeline_files_valhall/functions/fu_context_files_oid_fileshare_annotation/pipeline.py [hunk 8]
```diff
@@ -0,0 +1,554 @@
...
+        dict[str, int]: coordinates used by annotations.
+    """
+    x_min, *_, x_max = sorted(min(1, vert["x"]) for vert in vertices)
+    y_min, *_, y_max = sorted(min(1, vert["y"]) for vert in vertices)
+
+    # Adjust if min and max are equal
+    if x_min == x_max:
+        x_min, x_max = x_min - 0.001, x_max if x_min > 0.001 else x_min, x_max + 0.001
+    if y_min == y_max:
+        y_min, y_max = y_min - 0.001, y_max if y_min > 0.001 else y_min, y_max + 0.001
```
Comment (doctrino) [inline cognite_toolkit/cognite_modules/examples/cdf_data_pipeline_files_valhall/functions/fu_context_files_oid_fileshare_annotation/pipeline.py:515]:
```suggestion
    if x_min == x_max:
        x_min, x_max = (x_min - 0.001, x_max) if x_min > 0.001 else (x_min, x_max + 0.001)
    if y_min == y_max:
        y_min, y_max = (y_min - 0.001, y_max) if y_min > 0.001 else (y_min, y_max + 0.001)
```
If not you are unpacking 3 values, i.e., `x_min-0.001, (x_max if x_min > 0.001 else x_min), x_max +0.001`

File: cognite_toolkit/cognite_modules/examples/cdf_data_pipeline_files_valhall/functions/fu_context_files_oid_fileshare_annotation/pipeline.py [hunk 9]
```diff
@@ -0,0 +1,554 @@
...
+                asset_ids_list = asset_ids_list[:1000]
+
+            if config.debug:
+                print(f"[INFO] Converted and created (not upload due to DEBUG) file: {file_xid}")
+                print(f"[INFO] Assets found: {asset_names}")
+                continue
+
+            annotated_count += 1
+            # Note: add a minute to make sure annotation time is larger than last update time:
+            timestamp = datetime.now(timezone.utc) + timedelta(minutes=1).strftime(ISO_8601)
```
Comment (doctrino) [inline cognite_toolkit/cognite_modules/examples/cdf_data_pipeline_files_valhall/functions/fu_context_files_oid_fileshare_annotation/pipeline.py:338]:
```suggestion
            timestamp = datetime.now(timezone.utc) + timedelta(minutes=1)
```
you cannot add a string to a datetime

**Review by doctrino — COMMENTED** [review]
In addition to the few comments. 

A disadvantage of splitting into multiple files is that you can no longer run it locally due to relative imports. For debugging (and understanding) being able to run the function locally is valuable. Wdyt?

File: cognite_toolkit/cognite_modules/examples/cdf_data_pipeline_files_valhall/functions/fu_context_files_oid_fileshare_annotation/pipeline.py (hunk 9 above)
Comment (haakonvt) [reply cognite_toolkit/cognite_modules/examples/cdf_data_pipeline_files_valhall/functions/fu_context_files_oid_fileshare_annotation/pipeline.py:338]: Haha wtf was I thinking here 😂

Comment (haakonvt) [reply cognite_toolkit/cognite_modules/examples/cdf_data_pipeline_files_valhall/functions/fu_context_files_oid_fileshare_annotation/pipeline.py:338]: Ah, just forgot a pair of `()`

File: cognite_toolkit/cognite_modules/examples/cdf_data_pipeline_files_valhall/functions/fu_context_files_oid_fileshare_annotation/pipeline.py (hunk 8 above)
Comment (haakonvt) [reply cognite_toolkit/cognite_modules/examples/cdf_data_pipeline_files_valhall/functions/fu_context_files_oid_fileshare_annotation/pipeline.py:515]: 🤦‍♂️ , thanks for catching! Seems like my fav activity lately is deleting necessary brackets 😂

**Review by doctrino — APPROVED** [review]

**Pull request merged**